Add freelisting fibers API

ObsidianMinor opened the pull request:
This abstracts the fiber stack implementation into a separate `FiberStack` class, allowing stacks to be saved and freelisted via `FiberPool`.

kentonv commented on c++/src/kj/async-inl.h:
```diff
@@ -1095,7 +1114,33 @@ template <typename Func>
   typedef _::FixVoid<_::ReturnType<Func, WaitScope&>> ResultT;
 
-  Own<_::FiberBase> intermediate = kj::heap<_::Fiber<Func>>(stackSize, kj::fwd<Func>(func));
+  Own<_::FiberStack> stack = kj::heap<_::FiberStack>(stackSize);
```
Ideally, we want to keep the amount of code in a template function as minimal as possible, since the compiler generates a separate copy of this whole function for each instantiation (which, in this function's case, is each place where it is called, since the `Func` is different every time!).

Instead of adding all this stack-allocating code here, what if `FiberBase`'s constructor accepted `size_t stackSize, Maybe<FiberPool&>`, and then did the appropriate allocation depending on whether a FiberPool was provided?

I think this would also allow you to move `FiberStack`'s definition entirely to the `.c++` file, which is always nice since the header file needs to be re-parsed for every single dependent file we compile.

ObsidianMinor replied:
Or if it was a separate constructor, yeah I could see that.

kentonv commented on c++/src/kj/async-inl.h:
```diff
@@ -1095,7 +1114,33 @@ template <typename Func>
+  return _::maybeReduce(kj::mv(result), false);
+}
+
+struct FiberPool::Node final: public _::FiberStack {
```
In fact, I think the use of inheritance here may be overkill. I think you could have `FiberStack` itself include `Maybe<Own<FiberStack>> freelistNext;`, which would just remain null for stacks that aren't part of a freelist. It's not as elegant but the code would be a bit easier to follow. These classes are all private implementation details anyway.

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -853,14 +920,52 @@ FiberBase::FiberBase(size_t stackSizeParam, _::ExceptionOrValue& result)
 }
 
+void FiberStack::initialize(FiberBase& fiber) {
+  KJ_ASSERT(this->fiber == nullptr);
```
Nitpick: Use `KJ_REQUIRE` to check preconditions (failure = there is a bug in the caller). Use `KJ_ASSERT` is to test assumptions (failure = there's a bug in this code).

(In practice, they do exactly the same thing, but I like to use one or the other for self-documentation... admittedly I might eventually give up on this since even I mix them up fairly regularly.)

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -853,14 +920,52 @@ FiberBase::FiberBase(size_t stackSizeParam, _::ExceptionOrValue& result)
+
+FiberBase::FiberBase(size_t stackSize, _::ExceptionOrValue& result)
+    : state(WAITING), stack(kj::heap<FiberStack>(stackSize)), result(result) {
+    stack->initialize(*this);
```
Nitpick: too much indentation on this line.

jclee commented on c++/src/kj/async.c++:
```diff
@@ -291,31 +291,77 @@ Promise<void> TaskSet::onEmpty() {
+namespace _ {
+
+class FiberStack final {
+  // A class containing a fiber stack impl. This is seperate from fiber
```
s/seperate/separate

ObsidianMinor commented:
Ready for re-review. Added in some optimizations and fixed stacks so they can be reused properly. Test has also been updated to make sure fibers are reused.

ObsidianMinor commented:
Ping @kentonv

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -19,6 +19,8 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+#undef _FORTIFY_SOURCE // prevent longjmp from doing checks in fibers
```
Probably could use a clearer comment, like:

> If _FORTIFY_SOURCE is defined, longjmp() will complain when it detects the stack pointer is moving in the "wrong direction", thinking that you're jumping to a stack frame that no longer exists. However, we use longjmp() to jump between entirely different stacks to implement fibers. So, this debug check is not appropriate for us.

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -829,11 +835,13 @@ struct FiberStack::Impl {
+    impl.fiberContext.context.uc_stack.ss_size = allocSize - sizeof(Impl);
+    impl.fiberContext.context.uc_stack.ss_sp = reinterpret_cast<char*>(stack);
+    impl.fiberContext.context.uc_stack.ss_flags = 0;
+     // to reuse fibers the start fibers runs forever, so this doesn't
```
nitpick: Comment formatted weirdly. Extra space of indentation and wraps too early.

kentonv replied:
The comment is also a little obtuse, maybe we could phrase it as: "We don't use `uc_link` because our fiber start routine runs forever in a loop to allow for reuse. When we're done with the fiber, we just destroy it, without switching back to its stack. This is safe since the start routine doesn't allocate any memory or RAII objects before looping."

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -878,6 +886,11 @@ struct FiberStack::StartRoutine {
     ptr |= static_cast<uintptr_t>(static_cast<uint>(arg2)) << (sizeof(ptr) * 4);
     auto stack = reinterpret_cast<FiberStack*>(ptr);
 
+    /// Jump back into the fiber stack constructor
```
nitpick: three slashes?

ObsidianMinor replied:
The Rustdoc Way:tm: and a force of habit :)

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -878,6 +886,11 @@ struct FiberStack::StartRoutine {
     auto stack = reinterpret_cast<FiberStack*>(ptr);
 
+    /// Jump back into the fiber stack constructor
+    if (setjmp(&stack->impl.fiberContext.jumpBuf[0]) == 0) {
```
`&jumpBuf[0]` is redundant (kind of like saying `&*jumpBuf`). You should be able to write: `setjmp(stack->impl.originalContext.jumpBuf)`

Similarly for all the other setjmp/longjmp calls below.

ObsidianMinor replied:
Ah that makes a lot more sense than what the errors were telling me.

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -878,6 +886,11 @@ struct FiberStack::StartRoutine {
 
+    /// Jump back into the fiber stack constructor
+    if (setjmp(&stack->impl.fiberContext.jumpBuf[0]) == 0) {
+      swapcontext(&stack->impl.fiberContext.context, &stack->impl.originalContext.context);
```
This can be `setcontext()` rather than `swapcontext()` as we don't use `fiberContext.context` anymore after this point.

Though better yet, I think you could be using `longjmp()` here too to return to the calling fiber, if `FiberStack`'s constructor were to call `setjmp()`. And then you could use `setcontext()` rather than `swapcontext()` there.

Then you can go another step further: You only ever really need a `ucontext_t` when you are calling `makecontext()` to create the fiber initially. You could allocate that `ucontext_t` on the stack inside `FiberStack`'s constructor, rather than put it into `FastContext`. Then you could remove `FastContext::context` entirely; you only need `jumpBuf` there.

Let me know if that makes sense...

ObsidianMinor replied:
Set context stack info, make the context, setjmp, set the context to start, longjmp back. Got it.

ObsidianMinor commented:
Ping2 @kentonv

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -831,24 +830,26 @@ struct FiberStack::Impl {
-     // need to point to the main context
-    impl.fiberContext.context.uc_link = 0;
+    KJ_SYSCALL(getcontext(context));
+    context->uc_stack.ss_size = allocSize - sizeof(Impl);
```
Hmm, I think it would be clearer to move this initialization of `context` to happen inside `FiberStack`'s constructor, since it's no longer a member of `Impl`.

And then you don't have to change `impl` to be a pointer...

ObsidianMinor replied:
It'll require doing the pointer math for the stack allocation again inside the `FiberStack` ctor as well as `Impl::alloc`.

ObsidianMinor replied:
Done

kentonv replied:
You are right, the way you did it made more sense. Sorry about that. I guess you can remove the last commit.

Can you rebase the PR to make the commit history cleaner? I think it would make sense to squash down to two commits:

1. Implement freelist of fiber stacks.
2. Switch to setjmp/longjmp.

I think this should be just a matter of `git rebase -i master` and then squash the first sequence of commits and then the second sequence, and update the resulting commit descriptions -- probably no need to reorder anything.

ObsidianMinor commented:
Rebased and squashed @kentonv

kentonv commented on c++/src/kj/async.c++:
```diff
@@ -1047,15 +1056,20 @@ FiberStack::FiberStack(size_t stackSizeParam)
-  makecontext(&impl.fiberContext, reinterpret_cast<void(*)()>(&StartRoutine::run), 2, arg1, arg2);
+  makecontext(&context, reinterpret_cast<void(*)()>(&StartRoutine::run), 2, arg1, arg2);
 
+  if (setjmp(impl->originalJmpBuf) == 0) {
```
Oops, should be `_setjmp`?

kentonv replied:
Sorry for the run-around. Feel free to use `--amend` for this one fix, then should be good to merge.

ObsidianMinor replied:
Hmm how did that get there.

ObsidianMinor replied:
Fixed.
